### PR TITLE
SwiftUI Style .foo() wrappers

### PR DESCRIPTION
**Pull request by kyleve**
This PR implements SwiftUI-style wrappers which I think improve the succinctness and readability of Blueprint code.

The main file to check out for updates/examples to written consumer code is `ViewController.swift`.

### SwiftUI-Style `.foo()` methods for wrapping types
Before this change, if you wanted to have a scroll view, which wrapped a box, which wrapped an inset, which wrapped a constrained size, which wrapped a label (and so on), the code would look something like this:

```
ScrollView(.fittingHeight) (
   wrapping: Box(
        backgroundColor .lightGrey,
       wrapping: Inset(
          uniformInset: 10.0,
          wrapping: ConstrainedSize(
             height: .atLeast(20.0),
             wrapping: Label(
                text: "Hello, world!"
             )
         )
      )
   )
)
```

Which obviously is pretty verbose... It's nice that from the shape of the code you see what's going on, but it's pretty hard to follow in practice, especially if these inner statements get broken up into intermediary variables.

After this change, to write this code, this becomes like the following, similar to SwiftUI:

```
Label(text: "Hello, World!")
   .constrainedTo(height: .atLeast(20.0))
   .inset(by: 20.0)
   .box(background: .lightGrey)
   .scrollable(.fittingHeight)

```

### CODE REVIEW

File: BlueprintUI/Sources/Layout/ConstrainedAspectRatio.swift [hunk 1]
```diff
@@ -77,3 +77,17 @@ public struct ConstrainedAspectRatio: Element {
...
+
+
+public extension ConstrainedAspectRatio {
+    /// Initializes with the given properties.
+    ///
+    /// - parameters:
+    ///   - aspectRatio: The aspect ratio that the content size should match.
+    ///   - contentMode: Whether the aspect ratio should be reached by expanding the content
+    ///     element's size to fill its parent or shrinking it to fit.
+    ///   - wrapping: The content element.
```
Comment (kyleve) [inline BlueprintUI/Sources/Layout/ConstrainedAspectRatio.swift:89]: Remove

File: BlueprintUI/Sources/Layout/ConstrainedSize.swift [hunk 2]
```diff
@@ -51,6 +51,17 @@ extension ConstrainedSize {
 
 }
 
+public extension Element {
+    
+    func constrainedTo(
```
Comment (kyleve) [inline BlueprintUI/Sources/Layout/ConstrainedSize.swift:56]: Missing comment

File: BlueprintUI/Sources/Layout/Inset.swift [hunk 3]
```diff
@@ -63,13 +65,34 @@ public struct Inset: Element {
...
 
     public func backingViewDescription(bounds: CGRect, subtreeExtent: CGRect?) -> ViewDescription? {
         return nil
     }
+}
+
+
+public extension Element {
+    
+    func inset(top: CGFloat = 0.0, bottom: CGFloat = 0.0, left: CGFloat = 0.0, right: CGFloat = 0.0) -> Inset {
```
Comment (kyleve) [inline BlueprintUI/Sources/Layout/Inset.swift:81]: Missing comments here

File: BlueprintUI/Sources/Layout/Stack.swift [hunk 4]
```diff
@@ -7,9 +7,74 @@ import UIKit
 public protocol StackElement: Element {
     init()
     var layout: StackLayout { get }
-    var children: [(element: Element, traits: StackLayout.Traits, key: AnyHashable?)] { get set }
+    var children: [StackChild] { get set }
 }
 
+public struct StackChild {
```
Comment (kyleve) [inline BlueprintUI/Sources/Layout/Stack.swift:13]: Missing comments

File: BlueprintUICommonControls/Sources/AccessibilityBlocker.swift [hunk 5]
```diff
@@ -19,5 +19,10 @@ public struct AccessibilityBlocker: Element {
             config[\.accessibilityElementsHidden] = true
         }
     }
+}
 
+public extension Element {
+    func disableAccessibility() -> AccessibilityBlocker {
```
Comment (kyleve) [inline BlueprintUICommonControls/Sources/AccessibilityBlocker.swift:25]: `blockAccessibility()`

File: BlueprintUICommonControls/Sources/AccessibilityElement.swift [hunk 6]
```diff
@@ -105,5 +105,25 @@ public struct AccessibilityElement: Element {
             config[\.isAccessibilityElement] = true
         }
     }
+}
+
 
+public extension AccessibilityElement {
```
Comment (kyleve) [inline BlueprintUICommonControls/Sources/AccessibilityElement.swift:111]: Missing comment

File: BlueprintUICommonControls/Sources/Box.swift [hunk 7]
```diff
@@ -103,7 +114,26 @@ extension Box {
         case none
         case simple(radius: CGFloat, opacity: CGFloat, offset: CGSize, color: UIColor)
     }
+}
 
+public extension Element {
```
Comment (kyleve) [inline BlueprintUICommonControls/Sources/Box.swift:119]: Missing comment

**Comment by kylebshr** [comment]
**1) SwiftUI methods**

I really like this, and actually added similar functions in a side project over the weekend for some custom "view modifiers". I think it's both easier to read and write - I often start out with a base element like a label, and in the current form it's clunky to repeatedly wrap it in elements, and often "obscures" the real UI. 

**2) Stack improvements**

*StackChild*: I think this makes sense, having named static methods instead of having to read the docs for grow/shrink priority every time seems great (jokes aside, the names are way more readable and beginner friendly).

*Operators*: I'm not a huge fan of custom operators for UI, since they're not very discoverable, lead to divergence in code style (since it doesn't make sense to remove the functions) and it just doesn't "feel right" to me. Not a super strong opinion here though :)

**Comment by kyleve** [comment]
@kylebshr re operators: generally agree, but I think there's a time and place for them, eg common operations like this. As long as you can do everything _without_ operators that you can do with them, I feel it's fine.

**Comment by kyleve** [comment]
@kyleve: TODO: Split Stack changes out into a new PR

**Comment by kyleve** [comment]
I've removed the stack changes from this PR

File: SampleApp/Sources/XcodePreviewDemo.swift [hunk 8]
```diff
@@ -16,7 +16,7 @@ struct TestElement : ProxyElement {
             $0.verticalUnderflow = .justifyToStart
             
             for index in 1...12 {
-                $0.add(child: Label(text: "Hello, World") {
+                $0.add(growPriority: 0.0, shrinkPriority: 00, child: Label(text: "Hello, World") {
```
Comment (kyleve) [inline SampleApp/Sources/XcodePreviewDemo.swift]: Revert

File: BlueprintUI/Sources/Layout/Aligned.swift [hunk 9]
```diff
@@ -111,3 +111,23 @@ public struct Aligned: Element {
...
+
+public extension Element {
+    /// Wraps the element in an `Aligned` element with the provided parameters.
+    ///
+    /// - parameters:
+    ///   - vertically: The vertical alignment. Defaults to `.centered`.
+    ///   - horizontally: The horizontal alignment. Defaults to `.centered`.
+    ///
+    func aligned(
+        vertically: Aligned.VerticalAlignment = .center,
```
Comment (kylebshr) [inline BlueprintUI/Sources/Layout/Aligned.swift:123]: I don't love a default in this case because it's not clear reading the code that `.aligned()` means center-aligned. What do you think about removing the default and adding a `centered()` function if that's a common use case?

Edit: looks like there's already `centered`! Let me know what you think about removing the default

File: BlueprintUI/Sources/Layout/ConstrainedAspectRatio.swift [hunk 10]
```diff
@@ -77,3 +77,21 @@ public struct ConstrainedAspectRatio: Element {
...
+
+public extension ConstrainedAspectRatio {
+    /// Initializes with the given properties.
+    ///
+    /// - parameters:
+    ///   - aspectRatio: The aspect ratio that the content size should match.
+    ///   - contentMode: Whether the aspect ratio should be reached by expanding the content
+    ///     element's size to fill its parent or shrinking it to fit.
+    ///
+    func aspectRatio(
```
Comment (kylebshr) [inline BlueprintUI/Sources/Layout/ConstrainedAspectRatio.swift:90]: Would it be a good idea to always match the Layout name? I might type `.con..` expecting to find `ConstrainedAspectRatio` if I was already familiar with the type. No strong opinion though, there's an argument to be made for keeping them succinct

File: BlueprintUICommonControls/Sources/AccessibilityBlocker.swift [hunk 11]
```diff
@@ -19,5 +20,10 @@ public struct AccessibilityBlocker: Element {
             config[\.accessibilityElementsHidden] = true
         }
     }
+}
 
+public extension Element {
+    func blockAccessibility() -> AccessibilityBlocker {
```
Comment (kylebshr) [inline BlueprintUICommonControls/Sources/AccessibilityBlocker.swift:26]: Doc comment

File: BlueprintUICommonControls/Sources/AccessibilityElement.swift [hunk 12]
```diff
@@ -105,5 +105,25 @@ public struct AccessibilityElement: Element {
...
+    func accessibility(
```
Comment (kylebshr) [inline BlueprintUICommonControls/Sources/AccessibilityElement.swift:112]: Doc comment

File: BlueprintUICommonControls/Sources/ScrollView.swift [hunk 13]
```diff
@@ -61,6 +68,20 @@ public struct ScrollView: Element {
     }
 }
 
+public extension Element {
+    func scrollable(
```
Comment (kylebshr) [inline BlueprintUICommonControls/Sources/ScrollView.swift:72]: Doc comment

File: BlueprintUI/Sources/Layout/Aligned.swift (hunk 9 above)
Comment (kyleve) [reply BlueprintUI/Sources/Layout/Aligned.swift:123]: I kept the defaults from the normal init on `Aligned`, since sometimes you only want to pass one value, and I think centered is a pretty reasonable default. Thoughts?

File: BlueprintUI/Sources/Layout/ConstrainedAspectRatio.swift (hunk 10 above)
Comment (kyleve) [reply BlueprintUI/Sources/Layout/ConstrainedAspectRatio.swift:90]: Yeah, that's fair, will change – I shortened it since an "Aspect Ratio" is already inherently constrained, but adding `constrained` to the front is nice for symmetry with the type name and `ConstrainedSize` as well.

Comment (kyleve) [reply BlueprintUI/Sources/Layout/ConstrainedAspectRatio.swift:90]: Actually, `constrainedTo(aspectRatio:` works nicely as well, going with that.

File: BlueprintUI/Sources/Layout/ConstrainedAspectRatio.swift [hunk 14]
```diff
@@ -77,3 +77,21 @@ public struct ConstrainedAspectRatio: Element {
         }
     }
 }
+
+
+public extension ConstrainedAspectRatio {
```
Comment (kyleve) [inline BlueprintUI/Sources/Layout/ConstrainedAspectRatio.swift:82]: This should be on `Element`, doh.

File: BlueprintUI/Sources/Layout/Aligned.swift (hunk 9 above)
Comment (kyleve) [reply BlueprintUI/Sources/Layout/Aligned.swift:123]: Eh, what the hell, removing the default

File: BlueprintUICommonControls/Sources/AccessibilityElement.swift [hunk 15]
```diff
@@ -105,5 +105,28 @@ public struct AccessibilityElement: Element {
             config[\.isAccessibilityElement] = true
         }
     }
+}
+
 
+public extension AccessibilityElement {
```
Comment (kyleve) [inline BlueprintUICommonControls/Sources/AccessibilityElement.swift:111]: Element

**Comment by kylebshr** [comment]
Haha, didn’t realize the init had a default, I don’t like it there for the same reason but I’ll let you decide!

**Comment by mattfaluotico** [comment]
Popping in for a comment on Stack Improvements, even though this is merged. IMO you get a bit more clarity about what is being added when the Element proceeds the stacking traits. 

Was [dabbling](https://github.com/mattfaluotico/Blueprint/commit/11ed5f11c1bbf5c65361247b1e7432ef1a2c98ca) and added a similar StackChild, with the flex traits using function builders (like this PR adds)

so you could get:  

```
Row {
    Label(text: "A truly great label")
        .growPriority(1)
}
```

I like @kyleve's use of fixed a lot! 

```
Row {
    Label(text: "A truly great label")
        .fixed
    Spacer()
    Label(text: "Another great label")    
}
```
